File: src/main/java/application/TestGRPCClient.java
package application;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import kotlin.ranges.IntRange;

import java.util.stream.StreamSupport;

import static application.PingServiceGrpc.newBlockingStub;


public class TestGRPCClient {
    private final PingServiceGrpc.PingServiceBlockingStub stub;

    public TestGRPCClient(ManagedChannel channel) {
        this.stub = newBlockingStub(channel);
    }

    public void ping() {
        final Request request = Request.newBuilder().build();
        stub.ping(request);
    }

    public static void main(String[] args) {
        final ManagedChannel channel = ManagedChannelBuilder.forAddress("localhost", 8080).usePlaintext().build();
        final TestGRPCClient testGRPCClient = new TestGRPCClient(channel);

        long start = System.nanoTime();
        testGRPCClient.ping();
        long end = System.nanoTime();
        System.out.println("Connection: " + (end - start));

        start = System.nanoTime();
        testGRPCClient.ping();
        end = System.nanoTime();
        System.out.println("Before warm up: " + (end - start)) ;

        int n = 100_000;
        start = System.nanoTime();
        StreamSupport.stream(new IntRange(0, n).spliterator(), false)
                .forEach(integer -> testGRPCClient.ping());
        end = System.nanoTime();
        System.out.println("Warm up: " + (end - start) / n);

        start = System.nanoTime();
        testGRPCClient.ping();
        end = System.nanoTime();
        System.out.println("After warm up: " + (end - start));
    }
}

File: src/main/kotlin/application/TestGRPCClient.kt
package application

import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.runBlocking
import kotlin.system.measureNanoTime

class TestClient(private val channel: ManagedChannel) {
    private val stub: PingServiceGrpcKt.PingServiceCoroutineStub = PingServiceGrpcKt.PingServiceCoroutineStub(channel)

    suspend fun ping() {
        val request = Request.newBuilder().build()
        stub.ping(request)
    }
}

fun main() {
    val port = 8080
    val channel = ManagedChannelBuilder.forAddress("localhost", port).usePlaintext().build()
    val client = TestClient(channel)

    measureNanoTime {
        runBlocking {
            client.ping()
        }
    }.also {
        println("Connecting: $it")
    }

    measureNanoTime {
        runBlocking {
            client.ping()
        }
    }.also {
        println("Before warm up: $it")
    }

    val n = 100_000
    measureNanoTime {
        runBlocking {
            IntRange(0, n).map {
                async {
                    client.ping()
                }
            }.awaitAll()
        }
    }.also {
        println("Warm up: ${it / n}")
    }

    measureNanoTime {
        runBlocking {
            client.ping()
        }
    }.also {
        println("After warm up: $it")
    }
}